source/particle/interpolator/bilinear_least_squares.cc

```diff
@@ -110,11 +107,21 @@ namespace aspect
             const double particle_property_value = particle->get_properties()[property_index];
             r[index] = particle_property_value;
 
-            const Point<dim> particle_position = particle->get_location();
+            const Tensor<1, dim, double> relative_particle_position = (particle->get_location() - approximated_cell_midpoint) / cell_diameter;
             A(index,0) = 1;
-            A(index,1) = (particle_position[0] - approximated_cell_midpoint[0])/cell_diameter;
-            A(index,2) = (particle_position[1] - approximated_cell_midpoint[1])/cell_diameter;
-            A(index,3) = (particle_position[0] - approximated_cell_midpoint[0]) * (particle_position[1] - approximated_cell_midpoint[1])/std::pow(cell_diameter,2);
+            A(index, 1) = relative_particle_position[0];
+            A(index, 2) = relative_particle_position[1];
+            if (dim == 2)
+              {
+                A(index, 3) = relative_particle_position[0] * relative_particle_position[1];
+              }
+            else
+              {
+                A(index, 3) = relative_particle_position[2];
+                A(index, 4) = relative_particle_position[0] * relative_particle_position[1];
+                A(index, 5) = relative_particle_position[0] * relative_particle_position[2];
+                A(index, 6) = relative_particle_position[1] * relative_particle_position[2];
```

Comment:
I should have noticed this earlier, but what about the shape function index 7, i.e. `x*y*z`? Should that not be here as well?